Docs - add "Introduction to CDI" (#10529)
quarkusio/quarkus: docs/src/main/asciidoc/cdi.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Introduction to Contexts and Dependency Injection

include::./attributes.adoc[]
:numbered:
:sectnums:
:sectnumlevels: 4
:toc:

In this guide we're going to describe the basic principles of the programming model used all over the Quarkus universe. 
Quarkus programming model is based on the http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html[Contexts and Dependency Injection for Java 2.0, window="_blank"] specification.
However, a full CDI implementation is not used under the hood.
Quarkus only implements a subset of the CDI features - see also <<cdi-reference.adoc#supported_features,the list of supported features>> and <<cdi-reference.adoc#limitations,the list of limitations>>.
On the other hand, there are quite a few <<cdi-reference#nonstandard_features,non-standard features>> and <<cdi-reference.adoc#build_time_apis,Quarkus-specific APIs>>.

> We believe that our efforts will drive the innovation of the CDI specification towards the build-time oriented developer stacks in the future.

Reviewer: I think this part is too detailed for the introduction of this guide. I would put it at the end. Keep in mind that you target people not knowing CDI so they do not really care about this.

Author: OK, so move all sentences after `However,...`?

Reviewer 2: I think you need to choose between introduction to pure CDI where you would avoid Quarkus-specific examples (omitting no-args constructor for instance) and then you can move this text towards the end of the guide. Or you should state ahead that Quarkus does it slightly differently in the name of innovation and then we can keep the guide as-is. Just my 0.02$ :)

Author: > I think you need to choose between introduction to pure CDI...

I tend to disagree. IMHO it's ok to mix the contents. It's a quarkus guide anyway. I moved the sentences to the end.

== _Q: OK. Let's start simple. What is a bean?_

A: Well, a bean is a _container-managed_ object that supports a set of basic services, such as injection of dependencies, lifecycle callbacks and interceptors.

> == _Q: Wait a minute. What does it mean container-managed?_

Reviewer 3: ```suggestion
== _Q: Wait a minute. What does "container-managed" mean?_
```

A: Simply put, you don't control the lifecycle of the object instance directly.

> Instead, you can affect the lifecycle through the declarative means, such as annotations, configuration, etc.

Reviewer 3: ```suggestion
Instead, you can affect the lifecycle through declarative means, such as annotations, configuration, etc.
```

The container is the _environment_ where your application runs.
It creates and destroys the instances of beans, associates the instances with a designated context, and injects them into other beans.

== _Q: What is it good for?_

> A: An application developer can focus on the business logic rather than "where and how" to obtain a fully initialized component with all of its dependencies.

Reviewer 3: ```suggestion
A: An application developer can focus on the business logic rather than "where and how" of obtaining a fully initialized component with all of its dependencies.
```

Author: Hm, that does not sound good to me. What about _"An application developer can focus on the business logic **rather than finding out "where and how" to obtain** a fully initialized component with all of its dependencies."_?

Reviewer 3: Sure, that sounds good to me

NOTE: You've probably heard of the _inversion of control_ (IoC) programming principle. Dependency injection is one of the implementation techniques of IoC. 

> == _Q: OK. And what does a bean look like?_

Reviewer 3: ```suggestion
== _Q: But does a bean look like?_
```

Reviewer: There's a word missing in the suggestion.

Author: @geoand did you mean _"But what does a bean look like?"_. Maybe the simple _"What does a bean look like?"_ would be the best here...

Reviewer 3: > There's a word missing in the suggestion.

Of course there is... I wrote it :)

I indeed meant `But what does a bean look like?`.

A: There are several kinds of beans.
The most common ones are class-based beans:

.Simple Bean Example
[source,java]
----
import javax.inject.Inject;
import javax.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.metrics.annotation.Counted;

@ApplicationScoped <1>
public class Translator {

    @Inject
    Dictionary dictionary; <2>
    
    @Counted  <3>
    String translate(String sentence) { 
      // ...
    }
}
----

> <1> This is a scope annotation. It tells the container which context to associate the bean instance with. In this particular case, a *single bean instance* is created for the application and used by all dependants.

Reviewer 3: I am sure that users will now what `dependants` means here. Wouldn't it better to say `and used throughout the application`?

Reviewer 3: Yeah, that sounds fine aswell

Author: What about _"created for the application and used by all other beans that inject `Translator`."_? It's not so precise but it reads well...

> <2> This is a field injection point. It tells the container that `Translator` depends on the `Dictionary` bean. If there is no matching bean the build fails.

Reviewer: Yeah I would focus on the fact that the service will be made available to the bean via injection.

> <2> This is a field injection point. It tells the container that `Translator` depends on the `Dictionary` bean. If there is no matching bean the build fails.

Author: I'm sorry but I don't understand. It's all about "dependency injection" so if a component _depends_ on other component it must declare the dependency by means of an injection point. 

_"tells the container to set the field inside the bean, nothing more"_. Not really. The container must validate all dependencies/injection points first.

_"service will be made available to the bean via injection"_ is IMHO cumbersome.

> <2> This is a field injection point. It tells the container that `Translator` depends on the `Dictionary` bean. If there is no matching bean the build fails.

Reviewer 2: I suppose it boils down to how much detail you want in a beginner guide. But I get your point and on second thought, I'd keep it as is.

> <3> This is an interceptor binding annotation. In this case, the annotation comes from the MicroProfile Metrics. The relevant interceptor intercepts the invocation and updates the relevant metrics.

Reviewer: Yeah, I would introduce it later too.

> <3> This is an interceptor binding annotation. In this case, the annotation comes from the MicroProfile Metrics. The relevant interceptor intercepts the invocation and updates the relevant metrics.

Author: This example shows a typical bean (although simplified) and interceptors are very typical (metrics, transactions, fault tolerance, etc.). I will add a link to the "Interceptors" chapter.

[source,java]
----
public class Translator {

    @Inject
    Instance<Dictionary> dictionaries; <1>
    
    String translate(String sentence) { 
      for (Dictionary dict : dictionaries) { <2>
         // ...
      }
    }
}
----
<1> This injection point will not result in an ambiguous dependency even if there are multiple beans that implement the `Dictionary` type.
<2> `javax.enterprise.inject.Instance` extends `Iterable`.
====

== _Q: Can I use setter and constructor injection?_

A: Yes, you can.
In fact, in CDI the "setter injection" is superseded by more powerful https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#initializer_methods[initializer methods, window="_blank"].
Intializers may accept multiple parameters and don't have to follow the JavaBean naming conventions.

.Initialized and Constructor Injection Example
[source,java]
----
@ApplicationScoped
public class Translator {

    private final TranslatorHelper helper

    Translator(TranslatorHelper helper) { <1>
       this.helper = helper;
    }

    @Inject <2>
    void setDeps(Dictionary dic, LocalizationService locService) { <3>
      / ...
    }
}
----

> <1> This is a constructor injection. In fact, this code would not work in regular CDI implementations where a bean with a normal scope must always declare a no-args constructor. In Quarkus we detect this problem and "add" such a constructor directly in the bytecode.

Reviewer 3: Should it also be mentioned that no `@Inject` is needed here since there is only one constructor?

Reviewer: I dunno if we should mention the fact that it's not supported by other implementations here. Do we really care?

Reviewer 2: +1 to mention you don't need `@Inject`
And I think we can avoid the no-args comment instead, most users won't even know it's needed and the fact that it is required just rises more questions IMO

Author: > ...since there is only one constructor?

+1

> Do we really care?

We do care because similar features simplify the development a lot.

<2> An initializer method must be annotated with `@Inject`. 
<3> An initializer may accept multiple parameters - each one is an injection point.

> == _Q: You talked about some qualifiers?_

Reviewer 2: Should we also mention why would you want to use qualifiers? E.g. to distinguish between several beans having same type(s)?

Author: @manovotn isn't "Qualifiers are annotations that help the container to distinguish beans that implement the same type" clear enough?

A: https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#qualifiers[Qualifiers, window="_blank"] are annotations that help the container to distinguish beans that implement the same type.
As we already said a bean is assignable to an injection point if it has all the required qualifiers.
If you declare no qualifier at an injection point the `@Default` qualifier is assumed.

.Bean With Custom Qualifier Example
[source,java]
----
@Superior <1>
@ApplicationScoped
public class SuperiorTranslator extends Translator {

    String translate(String sentence) { 
      // ...
    }
}
----
<1> `@Superior` is a https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#defining_qualifier_types[qualifier annotation, window="_blank"].

> This bean would be assignable to `@Inject @Superior Translator` and `@Inject @Superior SuperiorTranslator` but not in `@Inject Translator`.

Reviewer: ```suggestion
This bean would be assignable to `@Inject @Superior Translator` and `@Inject @Superior SuperiorTranslator` but not to `@Inject Translator`.
```

The reason is that `@Inject Translator` is automatically transformed to `@Inject @Default Translator` during typesafe resolution.

> And since our `SuperiorTranslator` does not declare `@Default` only the original `Translator` bean is assignable.

Reviewer: Should we explain how you create a qualifier?

Author: +1

Reviewer 2: I think that's covered in the links to spec. You literally just copy the code it links to.

== _Q: Looks good. What is the bean scope?_

The scope of a bean determines the lifecycle of its instances, i.e. when and where an instance should be created and destroyed.

NOTE: Every bean has exactly one scope.

== _Q: What scopes can I actually use in my Quarkus application?_

A: You can use all the built-in scopes mentioned by the specification except for `javax.enterprise.context.ConversationScoped`.

[options="header",cols="1,1"]
|===
|Annotation   |Description   
//----------------------
|`@javax.enterprise.context.ApplicationScoped`   | A single bean instance is used for the application and shared among all injection points. The instance is created lazily, i.e. once a method is invoked upon the <<client_proxies, client proxy>>. 

> |`@javax.inject.Singleton`   | Just like `@ApplicationScoped` except that no client proxy is used. The instance is created eagerly.

Reviewer 2: The notion of being created eagerly can be misleading. It is eager as in not having to call a method on the bean. However, not as eager as automatically creating it on app startup (which is something we cover with another annotation). I'd probably remove that.

Author: _"The instance is created when an injection point that resolves to a `@Singleton` bean is being injected."_? I'd like to keep it because "no client proxy" info is not enough.

|`@javax.enterprise.context.RequestScoped`   | The bean instance is associated with the current _request_ (usually an HTTP request). 

> |`@javax.enterprise.context.Dependent`   | This is a pseudo-scope. There is no context object that holds all dependent instances. Instead, a dependent instace is a dependent object of the bean it is injected into.

Reviewer: ```suggestion
|`@javax.enterprise.context.Dependent`   | This is a pseudo-scope. There is no context object that holds all dependent instances. Instead, a dependent instance is a dependent object of the bean it is injected into.
```

> |`@javax.enterprise.context.Dependent`   | This is a pseudo-scope. There is no context object that holds all dependent instances. Instead, a dependent instace is a dependent object of the bean it is injected into.

Reviewer 2: I would be a bit more verbose, for a CDI newbie I think your description won't be clear enough (maybe I am not giving them enough credit though :)):
`This is a pseudo-scope. The instances are not shared and every injection point spawns a new instance of dependent bean. The lifecycle of dependent bean is bound to the bean injecting it - it will be created and destroyed along with the bean injecting it.`

> |`@javax.enterprise.context.SessionScoped`   | This scope is backed by an `javax.servlet.http.HttpSession` object. It's only available if `quarkus-undertow` extension is used.

Reviewer: ```suggestion
|`@javax.enterprise.context.SessionScoped`   | This scope is backed by an `javax.servlet.http.HttpSession` object. It's only available if the `quarkus-undertow` extension is used.
```

|===

> NOTE: There can be other custom scopes provided by Quarkus extension. For example, `qquarkus-narayana-jta` provides `javax.transaction.TransactionScoped`.

Reviewer 4: ```suggestion
NOTE: There can be other custom scopes provided by Quarkus extension. For example, `quarkus-narayana-jta` provides `javax.transaction.TransactionScoped`.
```

> NOTE: There can be other custom scopes provided by Quarkus extension. For example, `qquarkus-narayana-jta` provides `javax.transaction.TransactionScoped`.

Reviewer: ```suggestion
NOTE: There can be other custom scopes provided by Quarkus extensions. For example, `quarkus-narayana-jta` provides `javax.transaction.TransactionScoped`.
```

[[client_proxies]]
== _Q: I don't undestand the concept of client proxies._

Indeed, the https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#client_proxies[client proxies, window="_blank"] could be hard to grasp but they provide some useful functionality.
A client proxy is basically an object that delegates all method invocations to a target bean instance.
It's a container construct that implements `io.quarkus.arc.ClientProxy` and extends the bean class.

> The client proxy instance is always injected instead of a contextual instance of a normal scoped bean (e.g. `@ApplicationScoped` and `@RequestScoped`).

Reviewer 4: `contextual instance`: maybe find a more straightforward term: `instead of an object of the bean class`.

Reviewer 2: `contextual instance` is a term used within CDI spec. I think it is good to introduce it so that the reader understands it if they decide to open the specification.

Author: Hm, "contextual instance" is actually a well-defined term from the spec. "object of the bean class" is not always true because for intercepted beans it's actually an instance of a container construct (bean class subclass). I cann add a link to the spec plus show a generated client proxy pseudocode. As I said it's hard to find the balance ;-)

Client proxies allow for:

* Lazy instantiation - the instance is created once a method is invoked upon the proxy.
* Ability to inject a bean with "narrower" scope to a bean with "wider" scope; i.e. you can inject a `@RequestScoped` bean into an `@ApplicationScoped` bean.

> * Circular dependencies in the dependency graph. Circular dependencies are "bad design" but sometimes they're inevitable. 

Reviewer 3: ```suggestion
* Circular dependencies in the dependency graph. Having circular dependencies is often an indication that a redesign should be considered, but sometimes they're inevitable. 
```

* In rare cases it's practical to destroy the beans manually. A direct injected reference would lead to a stale bean instance.
 
 
== _Q: OK. You said that there are several kinds of beans?_

A: Yes. In general, in Quarkus we distinguish:

1. Class beans
2. Producer methods
3. Producer fields
4. Synthetic beans

Producer methods are useful if you need additional control over instantiation of a bean.
They are also useful when integrating third-party libraries where you don't control the class source and may not add additional annotations etc.

.Producer Method Example
[source,java]
----
@ApplicationScoped
public class PiProducer {

    @Produces <1>
    double producePi() { 
       return Math.PI; <2>

>     }

Reviewer 4: I'm not find of those examples that uses a CDI bean to produce String/long/doubles/...
This is not usually how we use them, we use them to instantiate and object from a third party library or to initialize complex objects.
Can you use a different example to make more sense ?

Author: Hm, I could redo the example to use some class from the JDK, e.g. `List` otherwise we would have to add one more class... Note that we want to demonstrate the principle and not a full-blown producer code.

Author: Actually, I can convert the PI producer to a producer field and add a producer method for a list... that makes a lot of sense IMO.

Reviewer 4: Maybe a `GreatingService` with a constructor that takes a greating String. This is a valid use case for a producer bean.
And it clearly demonstrate the use case of the creation of a _complex_ bean

}

@ApplicationScoped
public class PiConsumer {

   @Inject
   double pi; <3>
   
   double twoPis() {
     return pi * 2;
   }
} 
----

> <1> The container analyses the method annotations to build a bean metadata.

Reviewer 2: Maybe we should also note that a producer method needs to be declared inside another bean?

Author: I wouldn't complicate things here... in quarkus you don't even have to add a bean defining annotation to a class if it declares `@Produces`. I mean it's a good point but why not mention disposers then? Static producers? Etc. WDYT?

The return type is used to build the set of bean types. 
In this case, it will be `double` and `java.lang.Object`.
No scope annotation is declared and so it's defaulted to `@Dependent`.
<2> The container will delegate to this method when creating the bean instance.
<3> `PiProducer#producePi()` is invoked and the result is set to `PiConsumer#pi` when the `PiConsumer` instance is created.

NOTE: One of the limitations of producers is that interceptors are not applied to the result of a producer method/field.

> Producer fields are very similar to producer methods except that the field value is read when obtaining a bean instance.

Reviewer 3: Can we add a note that I personally hate producer fields? :laughing:

Reviewer: Do producer fields bring something to the mix? If not, maybe I wouldn't introduce them here.

Reviewer 2: > Do producer fields bring something to the mix? If not, maybe I wouldn't introduce them here.

They are just "simpler producer methods" because sometimes you don't need whole method to declare it.
The text related to producer fields is so tiny, I would probably keep it.

Author: > Can we add a note that I personally hate producer fields?

@geoand why is that? They're nice ;-).

Reviewer 3: They just don't fit the mental model I have in my head of how things are produced. Having something produced by calling a method makes sense, reading it from a field does not :)

TIP: There's more about producers. You can declare qualifiers, inject dependencies into the producer methods parameters, etc. You can read more about producers for example in the https://docs.jboss.org/weld/reference/latest/en-US/html/beanscdi.html#_producer_methods[Weld docs, window="_blank"].

Synthetic beans are usually provided by extensions.
A synthetic bean is not backed by a class/method/field.

> Application developers usually only consume this kind of beans.

Reviewer 4: As it didn't target developers, maybe don't introduce them ? It's and introduction guide we don't need to cover everything

Reviewer: Yeah, I think they shouldn't be introduced here.

Reviewer 2: I am biased about it. On one hand developers only consume them, OTOH they need to know they exist otherwise they will keep seeing injection of beans that have no classes/producers backing them and are seemingly conjured out of thin air.

Author: +1

Author: I only added a note to the list...

> == _Q: OK injection looks cool. What other services are provided?_

Reviewer 2: ```suggestion
== _Q: OK, injection looks cool. What other services are provided?_
```

=== Lifecycle Callbacks

A bean class may declare lifecycle `@PostConstruct` and `@PreDestroy` callbacks:

.Lifecycle Callbacks Example
[source,java]
----
import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;

@ApplicationScoped
public class Translator {

    @PostConstruct <1>
    void init() {
       // ...
    }
    
    @PreDestroy <2>
    void destroy() {
      // ...
    }
}
----
<1> This callback is invoked before the bean instance is put into service. It is safe to perform some initialization here.
<2> This callback is invoked before the bean instance is destroyed. It is safe to perform some cleanup tasks here.

TIP: It's a good practice to keep the logic in the callbacks "without side effects", i.e. you should avoid calling other beans inside the callbacks.

> 

Reviewer 4: Maybe add a note that Quarkus support custom callback global to the application via https://quarkus.io/guides/lifecycle ?

Author: Hm, there is no relation between lifecycle callbacks and `StartupEvent/@Startup`.

Reviewer 4: Right, one is bean lifecycle and the other application lifecycle. 
My suggestion was to add a note to distinguish between the two, but I'm OK to keep the doc as it is now.

=== Interceptors

Interceptors are used to separate cross-cutting concerns from business logic. 
There is a separate specification - Java Interceptors - that defines the basic programming model and semantics.

.Simple Interceptor Example
[source,java]
----
import javax.interceptor.Interceptor;
import javax.annotation.Priority;

@Logged <1>
@Priority(2020) <2>
@Interceptor <3>
public class LoggingInterceptor {

   @Inject <4>
   Logger logger;

   @AroundInvoke <5>
   Objec logInvocation(InvocationContext context) {
      // ...log before
      Objec ret = context.proceed(); <6>
      // ...log after
      return ret;
   }
    
}
----
<1> This is an interceptor binding annotation that is used to bind our interceptor to a bean. Simply annotate a bean class with `@Logged`.

> <2> `Priority` enables the interceptor and affects the the interceptor ordering. 

Reviewer 3: I think it makes sense to mention if the ordering is from highest to lowest or the inverse

Author: We could mention this but it's slightly more complicated because "Interceptors with smaller priority values are called first" but that actually means that interceptors with higher priority are "closer" to the business method invocation:

```
--> InterceptorWithPriority1 code before ctx.proceed()
   ---> InterceptorWithPriority10 code before ctx.proceed()
      ---> Business method
   ---> InterceptorWithPriority10 code after ctx.proceed()
--> InterceptorWithPriority1 code before ctx.proceed()
```
Wouldn't that confuse people even more? ;-)

Reviewer 3: I don't think it would confuse people. IMHO all I want to know as a user is which interceptor gets called first and which gets called last.

Author: Ok, no problem. Will add the _"Interceptors with smaller priority values are called first"_ sentence here. All confused people pls blame @geoand :D !

Reviewer 3: Send all the enraged users my way :)

> <2> `Priority` enables the interceptor and affects the the interceptor ordering. 

Reviewer 4: ```suggestion
<2> `Priority` enables the interceptor and affects the interceptor ordering. 
```

<3> Marks an interceptor component.
<4> An interceptor instance may be the target of dependency injection.

> <5> `AroundInvoke` denotes a method that interposes on business methods.

Reviewer 4: Maybe list the other interceptor pointcut ? (BeforeInvoke, AfterInvoke)

Author: I wouldn't complicate things here... BTW there is nothing like `BeforeInvoke` and `AfterInvoke` in the interceptors spec. There are `@AroundConstruct` and `@PostConstruct` and `@PreDestroy` interceptors...

<6> Proceed to the next interceptor in the interceptor chain or invoke the intercepted business method.

> NOTE: Instances of interceptors are dependent objects of the bean instance they intercept, i.e. a new interceptor instance is created for each intercepred bean.

Reviewer 4: ```suggestion
NOTE: Instances of interceptors are dependent objects of the bean instance they intercept, i.e. a new interceptor instance is created for each intercepted bean.
```

> === Events and Observers

Reviewer 4: Is event part of the CDI Spec ?
We already have the Vert.x eventbus for that purpose maybe we can skip this part or just link to the spec here.

Reviewer: I think it makes sense to have them here.

Author: Yes, events and observers are the core part of the CDI spec.

Reviewer 2: [Events and observers](https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#events) are part of core CDI. I don't think it's anyhow linked to Vert.x.

	
Beans may also produce and consume events to interact in a completely decoupled fashion.
Any Java object can serve as an event payload.
The optional qualifiers act as topic selectors.

.Simple Event Example
[source,java]
----

class TaskCompleted {
  // ...
}

@ApplicationScoped
class ComplicatedService {

   @Inject
   Event<Task> event; <1>

   void doSomething() {
      // ...
      event.fire(new TaskCompleted()); <2>
   }

}

@ApplicationScoped
class Logger {

>    void onTaskCompleted(@Observes TaskCompleted task) { <3>

Reviewer 2: I would add a second sample which will use qualifiers to demonstrate the typesafe resolution for observers.

Author: IMHO qualifiers are not used that often together with events/observers. I'd prefer to keep it simple. I will add a link to the weld docs though.

      // ...log the task
   }

}
----
<1> `javax.enterprise.event.Event` is used to fire events.
<2> Fire the event synchronously.
<3> This method is notified when a `TaskCompleted` event is fired.

== Conclusion

CDI is very mature and powerful technology.
In this guide, we've covered some of the basic topics.

> There is a Quarkus link:cdi-reference[CDI Reference Guide] that covers some Quarkus-specific features and limitations.

Reviewer 2: ```suggestion
If you wish to learn more, there is a Quarkus link:cdi-reference[CDI Reference Guide] that covers some Quarkus-specific features and limitations.
```

Author: Hm, what about _"If you wish to learn more about Quarkus-specific features and limitations there is a Quarkus link:cdi-reference[CDI Reference Guide]."_

> We also recommend you to read the http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html[CDI specification] and the https://docs.jboss.org/weld/reference/latest/en-US/html/[Weld documentation] (Weld is a CDI Reference Implementation) to get acquainted with more complex topics.

Reviewer 2: I am not sure linking to Weld makes sense here, I'd probably remove it.

Author: The Weld docs are a good source of info about CDI. I'm +10 to keep the link. BTW there are quite a few links to the Weld docs in the guide anyway....
